Add `database` field to `meta` for describing the current database (#424)
Materials-Consortia/OPTIMADE: optimade.rst

  Implementation note: the functionality of this field overlaps to some degree with features provided by the HTTP error :http-error:`429 Too Many Requests` and the `Retry-After HTTP header <https://tools.ietf.org/html/rfc7231.html#section-7.1.3>`__. Implementations are suggested to provide consistent handling of request overload through both mechanisms.

  - **database**: a dictionary describing the specific database accessible at this OPTIMADE API.

>     If provided, the dictionary fields SHOULD match those provided in the corresponding the links entry for the database in the provider's index meta-database, outlined in `Links Endpoint JSON Response Schema`_.

Reviewer: ```suggestion
    If provided, the dictionary fields SHOULD match those provided in the corresponding links entry for the database in the provider's index meta-database, outlined in `Links Endpoint JSON Response Schema`_.
```
I think this the is not necessary.

>     The dictionary should contain the OPTIONAL fields:

Reviewer: To me, it sounded a bit conflicting that you use both should and OPTIONAL here. 
If all the properties of the database dictionary are optional "can" would perhaps be more fitting.

>     - **name**: a human-readable name for the database, e.g., for use in clients.
>     - **description**: a human-readable description of the database, e.g., for use in clients.

Reviewer 2: ```suggestion
    - **name**: a human-readable name for the database, e.g., for use in clients.
    - **version**: a string describing the version of the database.
    - **description**: a human-readable description of the database, e.g., for use in clients.
```

    - **homepage**: a `JSON API links object <http://jsonapi.org/format/1.0/#document-links>`__, pointing to a homepage for the particular database.

>     - **maintainer**: a dictionary providing details about the maintainer of the database, which MUST contain the single field:

Reviewer: Sorry I did not spot this the first time, but what is the meaning of the word "single" in this line.
"- **maintainer**: a dictionary providing details about the maintainer of the database, which MUST contain the single field: "
Does this imply that the database can not add extra database specific fields here? If so, it does not seem necessary to have a separate dictionary for the maintainer. If it does not, I think it would be better to remove the word single here.

Author: I used same wording as `implementation->maintainer` but agree it is a bit clunky.

Reviewer: I think I would prefer if it is removed there as well.
@rartino do you know why you originally chose this wording?

Reviewer 2: Maybe bad form from my part to dig through the git blame log to reflect blame :-), but this phrasing seems to originate from @merkys ( https://github.com/Materials-Consortia/OPTIMADE/blame/ddd33aca159c22c7c8760bbd3aa2381164056d61/optimade.md#L250 ). I think it was only meant as a helper word to clarify that the reader should only expect a single field below. I do not see the harm in it, but also do not object to it being removed.

Reviewer: I did a git blame from my terminal and there it listed you as the author. But perhaps you merely moved some lines around. 
You definitely looked at an older version of the text than me.

Perhaps I am also adding too much importance to it.  "single" just seemed an unnecessary word in this sentence to me.

Reviewer 2: About the formulation: even with my most strict reading, I just see it saying "this one specific single field is required" with no implications for other, e.g., data-base specific, fields 

About git blame: you have to dig backwards in the git blame log (on GitHub, use the "stack of papers" icon next to the line number) to find who originally made the contribution. Otherwise I'm going to be blamed for almost everything due to the PR where I reformatted the specification from .md to .rst, and a few others where I moved whole sections.

Reviewer 3: This indeed might be me to blame :) I added this construction with the aim to forbid empty `maintainer` dictionaries. At the time of writing there could be only one field in `maintainer`, thus I used this wording. Also at that time I was not aware that prefixed database-specific fields were allowed in every dictionary (or did we change that a bit later?).

  Implementation note: the functionality of this field overlaps to some degree with features provided by the HTTP error :http-error:`429 Too Many Requests` and the `Retry-After HTTP header <https://tools.ietf.org/html/rfc7231.html#section-7.1.3>`__. Implementations are suggested to provide consistent handling of request overload through both mechanisms.

>   - **database**: a dictionary describing the specific database accessible at this OPTIMADE API.

Reviewer: Could you perhaps use in the word OPTIONAL in the description of the database field. That would be more consistent with how we describe optional fields elsewhere in the database.
```suggestion
  - **database**: an OPTIONAL dictionary describing the specific database accessible at this OPTIMADE API.
```

Reviewer 2: Isn't this placed under "meta MAY also include these fields"?

Reviewer: Yes you are right, I had not scrolled up far enough.

  Implementation note: the functionality of this field overlaps to some degree with features provided by the HTTP error :http-error:`429 Too Many Requests` and the `Retry-After HTTP header <https://tools.ietf.org/html/rfc7231.html#section-7.1.3>`__. Implementations are suggested to provide consistent handling of request overload through both mechanisms.

  - **database**: a dictionary describing the specific database accessible at this OPTIMADE API.
    If provided, the dictionary fields SHOULD match those provided in the corresponding links entry for the database in the provider's index meta-database, outlined in `Links Endpoint JSON Response Schema`_.
    The dictionary can contain the OPTIONAL fields:

>     - **name**: a human-readable name for the database, e.g., for use in clients.

Author: ```suggestion
    - **id**: the identifier of this database within those served by this provider, i.e., the ID under which this database is served in this provider's index meta-database.
    - **name**: a human-readable name for the database, e.g., for use in clients.
```

  Implementation note: the functionality of this field overlaps to some degree with features provided by the HTTP error :http-error:`429 Too Many Requests` and the `Retry-After HTTP header <https://tools.ietf.org/html/rfc7231.html#section-7.1.3>`__. Implementations are suggested to provide consistent handling of request overload through both mechanisms.

  - **database**: a dictionary describing the specific database accessible at this OPTIMADE API.
    If provided, the dictionary fields SHOULD match those provided in the corresponding links entry for the database in the provider's index meta-database, outlined in `Links Endpoint JSON Response Schema`_.
    The dictionary can contain the OPTIONAL fields:

    - **id**: the identifier of this database within those served by this provider, i.e., the ID under which this database is served in this provider's index meta-database.
    - **name**: a human-readable name for the database, e.g., for use in clients.
    - **version**: a string describing the version of the database.
    - **description**: a human-readable description of the database, e.g., for use in clients.

>     - **homepage**: a `JSON API links object <http://jsonapi.org/format/1.0/#document-links>`__, pointing to a homepage for the particular database.

Reviewer 4: The `homepage` should probably also be allowed to be as string, similarly to the `homepage` field under the `provider`. Note, that the currently provided example already uses it as a string field.

Reviewer 2: Oh, wait - does it make sense for these things (here and in multiple other places in the specification) to be "links objects", which allow multiple links with different keys? Or, did we actually mean to say "a link as defined in the JSON API specification under links objects http://jsonapi.org/format/1.0/#document-links ?

Author: Yes @rartino, I think you are right... Maybe I can suggest this in a separate PR

Author: https://github.com/Materials-Consortia/OPTIMADE/pull/474#issuecomment-1591806840

Reviewer 2: You're right, that first "(a “links object”)" parenthesis makes no sense to me either. Everything else makes sense as:

- "links object": an object where each member is a "link".
-  "link": a string or "link object".
- "link object": an object with a href member; in v1.1 it has its own subsection: https://jsonapi.org/format/1.1/#auto-id--link-objects

So, when we are trying to represent single links in our API, I'd say our options in increasing "JSON API-iness" (and perhaps decreasing OPTIMADE-iness") are:

1. "A string" equal to the URL.
2. "A link as described in http://jsonapi.org/format/1.1/#document-links" (i.e., string or "link object").
3. "A link object" as described in https://jsonapi.org/format/1.1/#auto-id--link-objects (must be an object with a href memeber, not allowed to be a string).
4. Move the link over to the closest JSON API-mandated links object, and define it there with a key that makes sense. For this PR it would be the topmost link under a key named "database").

Not sure what to do about this. Given how non-JSON-API-formatted this response already is, "1" seems attractive...

Reviewer 2: Seeing you https://github.com/Materials-Consortia/OPTIMADE/pull/474, I suspect you mean to do this before merging then?
```suggestion
    - **homepage**: a `JSON API link <http://jsonapi.org/format/1.0/#document-links>`__, pointing to a homepage for the particular database.
```
(Or do you want to merge this with the bug first, and then correct it in #474?)

Author: Sorry, this one got lost in the notifications... I've committed your suggestion, I think we're at least self-consistent now...

Author: Perhaps @rartino can re-review and I'll just merge and stop bothering everyone else? (unless they also review before I get the chance to merge...)

  - Example for a request made to :query-url:`http://example.com/optimade/v1/structures/?filter=a=1 AND b=2`:

    .. code:: jsonc

>       

Author: ```suggestion
```

               "email": "admin@example.com"
             },
             "issue_tracker": "http://tracker.example.com/exmpl-optimade"
           },

>            "database": {

Author: ```suggestion
           "database": {
               "id": "exmpl",
```

Reviewer: Would it perhaps be better to use a string other than "exmpl", as we also use that for the provider prefix?

Author: Fair, just changed it to "example"

Reviewer: Perhaps "example_db" would be even clearer, although I guess I should have suggested that in my first comment.

               "email": "admin@example.com"
             },
             "issue_tracker": "http://tracker.example.com/exmpl-optimade"
           },
           "database": {

>                "id": "example_db",

Reviewer 4: ```suggestion
             "id": "example_db",
```

  Implementation note: the functionality of this field overlaps to some degree with features provided by the HTTP error :http-error:`429 Too Many Requests` and the `Retry-After HTTP header <https://tools.ietf.org/html/rfc7231.html#section-7.1.3>`__. Implementations are suggested to provide consistent handling of request overload through both mechanisms.

  - **database**: a dictionary describing the specific database accessible at this OPTIMADE API.
    If provided, the dictionary fields SHOULD match those provided in the corresponding links entry for the database in the provider's index meta-database, outlined in `Links Endpoint JSON Response Schema`_.

>     The dictionary can contain the OPTIONAL fields:

Reviewer 4: ```suggestion
    The dictionary can contain the following OPTIONAL fields:
```

Author: I think to avoid another review round-trip I'll merge this as is; this can be added to a pre-release "tweaks" PR though.

Reviewer 4: Sure thing.
